src/AddAtomicMutex.cpp

```diff
@@ -0,0 +1,266 @@
+        }
+    }
+
+    std::set<string> remove_mutex_lock_names;
+};
+
+class FindAtomicMutexUsage : public IRGraphVisitor {
+public:
+    FindAtomicMutexUsage(const std::string &producer_name)
+        : producer_name(producer_name) {}
+
+    using IRGraphVisitor::visit;
+
+    void visit(const Atomic *op) override {
+        for (Expr i : op->mutex_indices) {
+            include(i);
+        }
+        include(op->body);
+        if (op->producer_name == producer_name &&
+                op->mutex_name != "") {
+            if (found) {
+                // Multiple atomics inside the producer,
+                // make sure they have consistent information
+                internal_assert(mutex_name == op->mutex_name &&
+                    tuple_size == op->tuple_size &&
+                    dimensions == op->dimensions) <<
+                    "Inconsistent informatino of atomics inside a Producer node.\n";
+            }
+            found = true;
+            mutex_name = op->mutex_name;
+            tuple_size = op->tuple_size;
+            dimensions = op->dimensions;
+        }
+    }
+
+    const std::string &producer_name;
+    bool found = false;
+    std::string mutex_name;
+    int tuple_size = 0;
+    int dimensions = 0;
+};
+
+class AddAtomicMutex : public IRMutator {
+public:
+    using IRMutator::visit;
+
+    const Target &target;
+
+    AddAtomicMutex(const Target &target) : target(target) {}
+
+    Stmt visit(const ProducerConsumer *op) override {
+        FindAtomicMutexUsage finder(op->name);
+        if (op->is_producer) {
+            finder.visit(op);
+            if (!finder.found) {
+                return IRMutator::visit(op);
+            }
+        } else {
+            return IRMutator::visit(op);
+        }
+
+        const std::string &producer_name = finder.producer_name;
+        const std::string &mutex_name = finder.mutex_name;
+        int tuple_size = finder.tuple_size;
+        int dimensions = finder.dimensions;
+        const char *extent_field = tuple_size == 1 ? ".extent." : ".0.extent.";
+        const char *min_field = tuple_size == 1 ? ".min." : ".0.min.";
+        const char *stride_field = tuple_size == 1 ? ".stride." : ".0.stride.";
+        Expr buffer_size = Expr(1);
+        for (int i = 0; i < dimensions; i++) {
+            Expr extent = Variable::make(Int(32), producer_name + extent_field + std::to_string(i));
+            buffer_size *= extent;
+        }
+        Expr mutex_array = Call::make(type_of<halide_mutex_array **>(),
+                                      "halide_mutex_array_create",
+                                      {buffer_size},
+                                      Call::Extern);
+        Stmt body = mutate(op->body);
+        // Allocate a scalar of halide_mutex_array.
+        // This generate halide_mutex_array *mutex[1];
+        body = Allocate::make(mutex_name,
+                              type_of<halide_mutex_array *>(),
```

Comment:
On second thought, the double-pointer may be necessary, because I don't think you can make an Allocate of `type_of<halide_mutex_array>()`

Reply:
I made the type a Handle() and it was fine. The codegen doesn't seem to check the type of the buffers.